### ISSUE TITLE
Anax kills shared container when it is still being used

### ISSUE THREAD

**Issue by bmpotter** [open-horizon/anax#178]
With the shared/singleton workload container support, 2 tests failed:

1. When the location and gpstest workloads were both running and stable, i manually docker stopped the location container. Anax detected that and stopped the gps container (which it shouldn’t have, because gpstest was still using the gps container). This caused the gpstest container to be restarted by anax also (because 1 of its containers wasn't running).
1. After the containers stabilized again, i deleted the gpstest agreement and anax brought both the gpstest and gps containers down (even though the location container was using the gps container). By the time it was bringing the gpstest workload back up (both gpstest and gps containers), the location container came down. (I think anax must have stopped it because the location log file didn’t indicate it had crashed or stopped itself). After that the location container was started and things stabilized.

The agbot location2-amd64.policy file deployment string is:

```
"{\"services\":{\"gps\":{\"image\":\"summit.hovitos.engineering/x86/gps:2.0.3\",\"privileged\":true,\"devices\":[\"/dev/bus/usb/001/001:/dev/bus/usb/001/001\"]},\"location\":{\"image\":\"summit.hovitos.engineering/x86/location:2.0.6\",\"environment\":[\"USE_NEW_STAGING_URL=false\"]}},\"service_pattern\":{\"shared\":{\"singleton\":[\"gps\"]}}}"
```

The gpstest2-amd64.policy file deployment string is:

```
"{\"services\":{\"gps\":{\"image\":\"summit.hovitos.engineering/x86/gps:2.0.3\",\"privileged\":true,\"devices\":[\"/dev/bus/usb/001/001:/dev/bus/usb/001/001\"]},\"gpstest\":{\"image\":\"summit.hovitos.engineering/x86/gps-test:latest\",\"environment\":[\"REPORTING_INTERVAL=1\",\"INTERVAL_SLEEP=5\",\"HEARTBEAT_TO_MQTT=true\"]}},\"service_pattern\":{\"shared\":{\"singleton\":[\"gps\"]}}}"
```

**Comment by dabooz** [comment]
Tried to recreate this today in order to debug it but am unable to recreate. The singleton gps container is staying up. Tried stopping containers, deleting agreements, but cannot get the system to kill the shared container.

**Comment by dabooz** [comment]
@bmpotter I think I will need you to recreate this issue with anax logging set to level 5 and grab the anax log after the system stabilizes. I am just not able to recreate this in my test environment.

**Comment by dabooz** [comment]
@bmpotter assign this back to me when you have recreated it. thanks

**Comment by bmpotter** [comment]
I recreate the 2 problems in the initial comment of this issue, exactly as described there. I'm doing this on my x86 device on staging, using the policy files location2-amd64.policy and gpstest2-amd64.policy. Assigning back to you to debug.

**Comment by dabooz** [comment]
Had to uplevel my docker to 17.05 in order to recreate the issue. Once I did, I found 2 other problems while testing that weren't seen before, so here is what I fixed:
1) when manually stopping a workload container, any shared containers that are part of that workload are not stopped unless that workload is the only user of the shared container.
2) when manually stopping a shared container, all agreements using that shared container are cancelled and recreated. The stopped shared container is removed and recreated/restarted.
3) when anax initializes and syncs up is containers and networks with existing agreements, shared container networks are removed if there are containers running on it.
